### ISSUE TITLE
Impact of `copy_as_padded`

### ISSUE THREAD

**Issue by art-w** [ocaml-multicore/saturn#171]
There are a bunch of places in the datastructures which makes use of `Multicore_magic.copy_as_padded`. For example:

https://github.com/ocaml-multicore/saturn/blame/fc34768b90a9f5503b19d1605cf115c241f51888/src/ws_deque.ml#L48

It's very hard to evaluate in the abstract which `copy_as_padded` are required (so I've to trust that intensive benchmarks were done) and I sometimes wonder if some are there "just in case". Is it possible that abusing `copy_as_padded` could negatively impact large program performances due to the increase memory usage? (even though the optim looks great in micro-benchmarks? or is the extra space small enough not to be a worry?)

Specifically, I'm wondering if the `copy_as_padded` that are done on the "root" record of datastructures (as in the link above) are 100% necessary to enable by default? Savvy users and micro-benchmarks can call `copy_as_padded` themselves in that case?

Also, it would be fun to automate the benchmarking / testing of which `copy_as_padded` are critical by removing one at a time and checking for an expected perf regression :)

**Comment by polytypic** [comment]
> Is it possible that abusing copy_as_padded could negatively impact large program performances due to the increase memory usage?

Yes, it will increase memory usage and increase the amount of work for GC.  So, for example, padding every node of a linked data structure typically reduces performance as such nodes are typically short lived and not (all of them) accessed frequently and so the reduction in false sharing does not outweigh the costs from the increased memory usage.

> Specifically, I'm wondering if the copy_as_padded that are done on the "root" record of datastructures (as in the link above) are 100% necessary to enable by default?

The root record of a data structure is just as vulnerable to false sharing as anything else.   And it is the most persistent and long-lived part of a data structure. 

People often seem to have wrong intuition and think that the CPU and the caches somehow distinguish between "atomic" and "non-atomic" locations in memory.  But, in reality, for the CPU and caches there is no such distinction at all.  Any cache line that happens to be frequently written to by some core (atomically or non-atomically) and is frequently read by multiple cores (atomically or non-atomically) quickly becomes a bottleneck and this can easily happen with the root node.  All it takes is to have some frequently enough mutated thing allocated next to the root.

Of course, the problem with false sharing is that sometimes you get lucky.  You remove padding and performance is the same.  You then conclude that it is not needed.  Later you make some seemingly unrelated change to the code and performance drops (and you might not notice this, because your changes are not related and so you are not careful), because now something mutable happens to get allocated or lated moved/compacted to be in the same cache line with some hot data structure root.

> Savvy users and micro-benchmarks can call copy_as_padded themselves in that case? (edit: but a real program that's not hammering the datastructure may not need it)

The way I've done this is that I've used [an optional `?padded:bool` argument](https://github.com/ocaml-multicore/picos/blob/84d54282d5b1cb0e80f51bf2bc05d30fec06ca65/lib/picos_aux.mpscq/picos_aux_mpscq.mli#L11) and [use `copy_as ?padded` in the implementation](https://github.com/ocaml-multicore/picos/blob/84d54282d5b1cb0e80f51bf2bc05d30fec06ca65/lib/picos_aux.mpscq/picos_aux_mpscq.ml#L16-L19).  I'm not sure it is the best approach, because in many use case you then need to pass `~padded:true` or you [risk having bad luck](https://www.youtube.com/watch?v=KALhynXxKp0).  It might be better to pad by default and then the lucky ones would have to specify `~feeling_lucky:true`.
